### DISCUSSION TITLE
Building a single bottle for all platforms

### DISCUSSION THREAD

**Discussion by Idane** [Homebrew/discussions#4639] [Tap maintenance and Homebrew development]
### Output of `brew config`

```shell
Irrelevant
```

### Output of `brew doctor`

```shell
Irrelevant
```

### Description of issue

Hi, I intend to distribute a jar with brew. The jar is identical for all platforms. I set up a tap here: https://github.com/krud-dev/homebrew-tap and used the default actions to generate the bottles. While experimenting I found that a bottle tagged with 'all' will work for all platforms.

My question is, how can I instruct test-bot to generate the bottle artifact as "all" rather than "x86_64_linux"?

**Comment by SMillerDev** [comment]
If it detects that the hashes for all platforms are the same it'll make it an `all` bottle.

**Reply by Idane** [reply]
I'm currently building for one platform to avoid unnecessary builds and it appears as "x86_64_linux", at what point is that condition satisfied? Any way to trigger it manually?

**Reply by SMillerDev** [reply]
I'd recommend just testing the compatibility on both platforms and let brew calculate the bottle type. I don't know of any way to force it.

**Reply by Idane** [reply]
Problem with jar files is that the hash will be different every time, so that won't solve it. Would you be open to a PR to add a flag to override this behaviour and force all?

**Reply by SMillerDev** [reply]
That seems like it won't be the same between OS versions then if it's not the same between runs, especially since Homebrew packages a bunch of jar files that do not have this issue.

**Reply by Idane** [reply]
I'm unsure of the difference between myself and these jars.
In any case, here is a simple project outlining this issue and showing that an identical build can produce two different hashes. I wrote a script that builds the same project in isolation twice, and prints out the hashes. You can take a look at the output [here](https://github.com/Idane/java-build-test/actions/runs/5543702734/jobs/10120090597) under `Build and Analyze`

```

First hash:  08b90f0a02fff853a9253a45873c6f681e67924811fdc8c45e9102f655b919ba build/first/libs/buildtestjava-1.0-SNAPSHOT.jar
Second hash:  7e55431714c83e236c3086724a44cf37bed63495f5ad477849d957f2676c97dc build/second/libs/buildtestjava-1.0-SNAPSHOT.jar
```

**Reply by gromgit** [reply]
It's been a while since I last Java'd, but if your JARs are anything like typical Java runtime ones, then it's either the manifest file that's different across platforms:
```
ubuntu$ unzip -p /usr/share/java/gettext.jar META-INF/MANIFEST.MF
Manifest-Version: 1.0
Created-By: 11.0.16 (Ubuntu)
```
or the order of files in the JAR, or their file timestamps.

In particular, zipping up a class hierarchy recursively is _not_ guaranteed to package individual classes in the same order _on the same platform each time_, much less on different platforms. This of course would _definitely_ affect the hash of the overall JAR, even if all the contents were identical.
